### PR TITLE
[Kernel][Misc] Add meta functions for ops to prevent graph breaks

### PR DESCRIPTION
**Pull request by bnellnm**
Miscellaneous changes to support `torch.compile` in vLLM.

- Add meta functions for various ops to prevent torch.compile graph breaks.
- Use string schemas for all dispatched op registrations.  (the function pointer API is only used for ops that are registered for all keys or that do not take Tensor arguments)
- Fix some type mismatches in the quantization code
- In the aqlm kernel/code, change codebook_partition_sizes into a list instead of a `Tensor` since it is allocated on the CPU.
- Bump up dynamo cache limits due to the amount of recompilation in cuda graph warmup.
- Add torch.library.opcheck tests for ops that had "unit" tests and are opcheck-able.

Note: opcheck does not seem to work with `torch.float8_e4m3fn`.  It complains that `mul_cuda` is not supported for that type.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by bnellnm** [comment]
/ready
/torch.compile

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -789,9 +793,10 @@ def load_model(self) -> None:
                     "This may lead to less accurate results!")
 
         if envs.VLLM_TEST_DYNAMO_GRAPH_CAPTURE:
-            self.model = torch.compile(self.model,
-                                       fullgraph=True,
-                                       backend="eager")
+            self.model = torch.compile(
+                self.model,
+                fullgraph=envs.VLLM_TEST_DYNAMO_FULLGRAPH_CAPTURE,
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:798]: can you explain the rationale here?

Comment (bnellnm) [reply vllm/worker/model_runner.py:798]: There's a different set of failures when `fullgraph=True`.   Not just graph breaks but actual dynamo errors.  Locally, I am switching between full/not-full constantly to test things.

File: vllm/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -18,6 +18,135 @@
 if TYPE_CHECKING:
     from vllm.worker.model_runner import ModelInputForGPUBuilder
 
+torch.library.define("vllm::flash_attn_varlen_func",
```
Comment (youkaichao) [inline vllm/attention/backends/flash_attn.py:21]: I believe `torch.library.define` is generally not necessary. We should use high-level APIs listed in https://pytorch.org/docs/main/library.html .

Comment (bnellnm) [reply vllm/attention/backends/flash_attn.py:21]: I tried using `custom_op` here but it couldn't deal with the window_size tuple so I ended up sticking with `define`

Comment (bnellnm) [reply vllm/attention/backends/flash_attn.py:21]: I think the longer term answer is to move the registration into the flash attn module itself once it is pulled into the main vllm build process.

File: vllm/worker/model_runner.py [hunk 3]
```diff
@@ -69,6 +69,10 @@
 
 TModelInputForGPU = TypeVar('TModelInputForGPU', bound="ModelInputForGPU")
 
+# For now, bump up cache limits for recompilations during CUDA graph warmups.
+torch._dynamo.config.cache_size_limit = 128
+torch._dynamo.config.accumulated_cache_size_limit = 128
+
```
Comment (youkaichao) [inline vllm/worker/model_runner.py:75]: I'm ok with this for now, but after I port the technique to skip dynamo guard evaluation overhead, this should not be necessary anymore.

File: vllm/model_executor/models/jamba.py [hunk 4]
```diff
@@ -731,7 +731,7 @@ def _clean_up_first_bs_blocks(self, batch_size: int,
                                   indices_for_current_run: List[int]):
         # move out all of the occupied but currently not running blocks
         # outside of the first n blocks
-        destination_indices = set([range(batch_size)])
+        destination_indices = range(batch_size)
```
Comment (youkaichao) [inline vllm/model_executor/models/jamba.py:734]: why do you need this?

File: vllm/distributed/parallel_state.py [hunk 5]
```diff
@@ -272,17 +272,17 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
 
         # Bypass the function if we are using only 1 GPU.
         if self.world_size == 1:
-            return input_
+            return input_.clone()
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:275]: this clone can hurt performance. why do we need this?

**Review by youkaichao — COMMENTED** [review]
thanks for the pr and sorry for keeping you waiting for so long!

the registration for the quantization kernels makes sense to me. but I don't understand the rest changes.

File: vllm/distributed/parallel_state.py (hunk 5 above)
Comment (bnellnm) [reply vllm/distributed/parallel_state.py:275]: This was part of the custom ar PR that didn't get removed.  I'll fix it.

**Comment by youkaichao** [comment]
feel free to ping me when it is ready for review again.

**Comment by bnellnm** [comment]
> feel free to ping me when it is ready for review again.

@youkaichao , it's ready for review again.

File: vllm/worker/model_runner.py (hunk 1 above)
Comment (youkaichao) [reply vllm/worker/model_runner.py:991]: can you send me the specific case where fullgraph fails?

Comment (bnellnm) [reply vllm/worker/model_runner.py:991]: There's probably more but these all fail with fullgraph=True and pass with fullgraph=False
```
FAILED tests/models/test_gguf.py::test_models[1-5-32-half-model0] - torch._dynamo.exc.Unsupported: hasattr TensorVariable shard_size
FAILED tests/models/test_gguf.py::test_models[1-5-32-half-model1] - torch._dynamo.exc.InternalTorchDynamoError: 'SymNodeVariable' object has no attribute 'value'
FAILED tests/models/test_gguf.py::test_models[1-5-32-half-model2] - torch._dynamo.exc.Unsupported: hasattr TensorVariable shard_size
FAILED tests/models/test_gguf.py::test_models[1-5-32-half-model3] - torch._dynamo.exc.Unsupported: hasattr TensorVariable shard_size
```

**Review by youkaichao — APPROVED** [review]
thanks for the great pr!

we can further discuss if we need to fix the graph breaks ourselves or pytorch team will fix them.

File: vllm/_custom_ops.py [hunk 6]
```diff
@@ -227,6 +243,194 @@ def gptq_marlin_24_gemm(a: torch.Tensor, b_q_weight: torch.Tensor,
                                             size_n, size_k)
 
 
+# TODO: has to be a better way to do this
+try:
+    torch.ops._C.gptq_marlin_24_gemm  # noqa B018
+
+    @torch.library.register_fake("_C::gptq_marlin_24_gemm")
```
Comment (gshtras) [inline vllm/_custom_ops.py:250]: This again breaks compatibility with torch < 2.4

**Pull request merged**